# ChatGPT Mental Support simulation

In this notebook we will simulate how chatGPT will respond to request from people with mental issues, problems and worries, that will be expressed plain in conversation.
This responses will be compared to responses from mental counselor to see how is chatGPT doing this.

### Dataset

This simulation is based on this dataset:

[Mental health Counseling Conversations dataset](https://www.kaggle.com/datasets/thedevastator/nlp-mental-health-conversations)

### Dataset structure

| Column   | Description | 
|----------|----------|
| Context  | Request, phrase, query sentence from the patient, that seems to have mental troubles, problem, disorders of different kind        |  
| Response | Response, that mental counselor is supposed to give. For one ```Context``` there is from 5 to 25 different responses | 


### Comparison

Comparison between ChatGPT response will be done with special prompt, resulting in score [0-100], and be done comparing one response from ChatGRT to all (one by one) responses from the dataset. New dataset will chatGPT responses and scores will be build, and then explored



## Preparing the data

The dataset on the link above is available to download, so here it is already present in the ```dataset/NLP_Mental_Health_Conversations.csv```

In [1]:
#!pip install pandas

In [2]:
import pandas as pd

C:\Users\sirius\AppData\Local\Temp\ipykernel_22444\4080736814.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [3]:
CONVERSATIONS_DATASET_FILENAME='dataset/NLP_Mental_Health_Conversations.csv'

conversation_dataset=pd.read_csv(CONVERSATIONS_DATASET_FILENAME)

conversation_dataset.head()

Context  \
0  I'm going through some things with my feelings...   
1  I'm going through some things with my feelings...   
2  I'm going through some things with my feelings...   
3  I'm going through some things with my feelings...   
4  I'm going through some things with my feelings...   

                                            Response  
0  If everyone thinks you're worthless, then mayb...  
1  Hello, and thank you for your question and see...  
2  First thing I'd suggest is getting the sleep y...  
3  Therapy is essential for those that are feelin...  
4  I first want to let you know that you are not ...

In [4]:
len(conversation_dataset)

3512

In [5]:
conversation_dataset.columns = conversation_dataset.columns.str.lower().str.replace(' ', '_')

In [6]:
conversation_dataset.isna().sum()

context     0
response    4
dtype: int64

In [7]:
conversation_dataset.describe()

context  \
count                                                3512   
unique                                                995   
top     I have so many issues to address. I have a his...   
freq                                                   94   

                                                 response  
count                                                3508  
unique                                               2479  
top     It's normal to feel a little anxiety--after al...  
freq                                                    3

In [8]:
context_example=conversation_dataset.loc[328]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)


I'm in a state of depression right now. Who can I talk to? I've been sick in a lot of pain and crying. Don't know where to turn.


In [9]:
responses_example

context  \
328   I'm in a state of depression right now. Who ca...   
2166  I'm in a state of depression right now. Who ca...   

                                               response  
328   First of all, if you’re sick and in pain, have...  
2166  First of all, if you’re sick and in pain, have...

In [10]:
responses_example.reset_index(inplace=True)
responses_example
print(responses_example.loc[0]['response'])
print(responses_example.loc[1]['response'])

First of all, if you’re sick and in pain, have you seen your doctor? If you have a chronic illness and pain, this can often lead to depression and is very common. You’ve taken the first step to dealing with it. You can talk to your doctor, even if your depression isn’t related to your health. He or she can either look at prescribing you an antidepressant or can refer you to a psychologist or psychiatrist.If that’s not really an option for you, there are also helplines (local or national) or support groups where people dealing with similar issues can talk to each other. If you can't find any in your area (newspapers usually publish this information), there are online support groups as well.
First of all, if you’re sick and in pain, have you seen your doctor? If you have a chronic illness and pain, this can often lead to depression and is very common. You’ve taken the first step to dealing with it. You can talk to your doctor, even if your depression isn’t related to your health. He or s

### Dataset contains duplicate responses

lets see some more

In [11]:
context_example=conversation_dataset.loc[38]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I have so many issues to address. I have a history of sexual abuse, I’m a breast cancer survivor and I am a lifetime insomniac.    I have a long history of depression and I’m beginning to have anxiety. I have low self esteem but I’ve been happily married for almost 35 years.
   I’ve never had counseling about any of this. Do I have too many issues to address in counseling?


In [12]:
responses_example

context  \
23    I have so many issues to address. I have a his...   
24    I have so many issues to address. I have a his...   
25    I have so many issues to address. I have a his...   
26    I have so many issues to address. I have a his...   
27    I have so many issues to address. I have a his...   
...                                                 ...   
2010  I have so many issues to address. I have a his...   
2011  I have so many issues to address. I have a his...   
2012  I have so many issues to address. I have a his...   
2013  I have so many issues to address. I have a his...   
2014  I have so many issues to address. I have a his...   

                                               response  
23    Let me start by saying there are never too man...  
24    It is never too late to get help and begin mak...  
25    You have been through so much and it sounds li...  
26    Absolutely not.  I strongly recommending worki...  
27    Absolutely not!  In fact, most people have man...  
...                                                 ...  
2010  The way it goes is that all your issues are co...  
2011  I think this is a very common question that pe...  
2012  Hello, I'm so glad you decided to take the fir...  
2013  Not at all my dear. Human beings are complex c...  
2014  Everyone has topics to discuss when they start...  

[94 rows x 2 columns]

In [13]:
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [14]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   
2701  I'm fine when we start becoming intimate, but ...   
2709  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...  
2701  Have you sought the support of a therapist rea...  
2709   Sexual intimacy can be very triggering for su...

### Removing duplicates and zero entries

In [15]:
conversation_dataset.dropna(subset=['response'], inplace=True)
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

In [16]:
len(conversation_dataset)

2479

In [17]:
conversation_dataset.isna().sum()

context     0
response    0
dtype: int64

In [18]:
context_example=conversation_dataset.loc[328]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)


I'm in a state of depression right now. Who can I talk to? I've been sick in a lot of pain and crying. Don't know where to turn.


In [19]:
responses_example

context  \
328   I'm in a state of depression right now. Who ca...   
2025  I'm in a state of depression right now. Who ca...   

                                               response  
328   First of all, if you’re sick and in pain, have...  
2025  First of all, if you’re sick and in pain, have...

In [20]:
conversation_dataset.describe()

context  \
count                                                2479   
unique                                                912   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2479  
unique                                               2479  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [21]:
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [22]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   
2221  I'm fine when we start becoming intimate, but ...   
2226  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...  
2221  Have you sought the support of a therapist rea...  
2226   Sexual intimacy can be very triggering for su...

In [23]:
conversation_dataset.loc[1000]['response']==conversation_dataset.loc[2226]['response']

False

In [24]:
print(conversation_dataset.loc[1000]['response'])
print(conversation_dataset.loc[2226]['response'])

print(conversation_dataset.loc[328]['response'])
print(conversation_dataset.loc[2025]['response'])

print(conversation_dataset.loc[328]['response']==conversation_dataset.loc[2025]['response'])

Sexual intimacy can be very triggering for survivors even when it is both wanted and consensual.  You may want to consider seeing a therapist who specializes in trauma to work through the abuse if you have not already done so. Often times triggers still hold such a powerful effect when the emotions related to the abuse  have not been fully processed.  In the meantime, you may want to consider coming up with a Safe Word to let your partner know that you are being triggered or to communicate your physical boundaries to him.  Often times, the experience of communicating  your physical boundaries to your partner, having those boundaries respected and validated, and having a partner who is understanding and  willing to engage in intimacy in such a way that does not violate your physical boundaries  can reinforce a sense of safety with him.
 Sexual intimacy can be very triggering for survivors even when it is both wanted and consensual.  You may want to consider seeing a therapist who specia

In [25]:
response1 = conversation_dataset.loc[328]['response']
response2 = conversation_dataset.loc[2025]['response']

# Find the minimum length of the two responses
min_length = min(len(response1), len(response2))

# Iterate through each character and find the positions where they differ
differences = []
for i in range(min_length):
    if response1[i] != response2[i]:
        differences.append(i)

# Print the positions where the responses differ
print("Positions where responses differ:")
print(differences)
print('Lengths of response1 and response2:', len(response1), len(response2))

Positions where responses differ:
[]
Lengths of response1 and response2: 698 699


In [26]:
print(response1.strip()==response2.strip())

True


So there is duplicate entries, but they differ in heading or trailing spaces.
Let's fix this, and remove duplicates once more.

In [27]:
conversation_dataset['response']=conversation_dataset['response'].apply(lambda x: x.strip())

In [28]:
conversation_dataset.describe()

context  \
count                                                2479   
unique                                                912   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2479  
unique                                               2075  
top     This can be a very complicated situation. The ...  
freq                                                    2

In [29]:
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

In [30]:
conversation_dataset.describe()

context  \
count                                                2075   
unique                                                847   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2075  
unique                                               2075  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [31]:
# Test once more
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [32]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...

### Errors in dataset

Working with the dataset I found some irrelevant entries. So we should just delete them


In [33]:
error_context=[" such as not enough sleep", 
"Can a counselor take sides with one parent and allow a parent to order the child to tell the counselor"]

conversation_dataset=conversation_dataset[~conversation_dataset['context'].str.startswith(tuple(error_context))]


In [34]:
conversation_dataset.describe()

context  \
count                                                2072   
unique                                                845   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2072  
unique                                               2072  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [35]:
context_unique = conversation_dataset.groupby('context').size().reset_index(name='count')

In [36]:
context_unique.sort_values('count',ascending=False)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
546  Is it normal for people to cry during therapy,...     38
524  I've gone to a couple therapy sessions so far ...     26
462  I'm going through some things with my feelings...     23
..                                                 ...    ...
361  I stress over everything. If I don't have enou...      1
366  I think I have depression, anxiety, bipolar di...      1
369  I think adult life is making him depressed and...      1
370  I think i may suffer from depression, and it i...      1
844  i was addicted to porn since the second grade....      1

[845 rows x 2 columns]

In [37]:
context_unique.sort_values('count',ascending=False).head(30)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
546  Is it normal for people to cry during therapy,...     38
524  I've gone to a couple therapy sessions so far ...     26
462  I'm going through some things with my feelings...     23
832  When my daughter is stressed about a silly thi...     23
80     How does a person start the counseling process?     22
31   Does counseling really do anything that can he...     21
79   How does a counselor decide when to end counse...     21
73   How do I decide if I should end my relationshi...     15
41   Every winter I find myself getting sad because...     15
699  My mother takes care of niece whom my sister a...     13
354  I start counseling/therapy in a few days (I'm ...     12
305  I just took a job that requires me to travel f...     11
311  I keep having these random thoughts that I don...     11
647  My fiancé and I have been together for 3 years...     11
827  When I'm in large crowds I get angry and I jus...     11
397  I was raped a couple months ago, Since then, a...     11
697  My mother has Alzheimer's and she has become s...     10
213  I have a friend that who I used to be in a rel...     10
485  I'm obsessing about a terrible breakup. Everyt...      9
817                What makes a healthy marriage last?      9
486  I'm planning to have baby, so I have to quit s...      9
568  I’ve been in a long distance relationship with...      9
154  I don't know what to say. I have never really ...      9
626  My daughter is in later elementary school. She...      9
563  I’m facing severe depression and anxiety and I...      9
16   After he got home from the hospital he was ang...      9
303  I just lost my grandpa and i'm having a rough ...      9
444  I'm being verbally abused on a daily basis by ...      9

In [38]:
context_unique['count'].describe()

count    845.000000
mean       2.452071
std        3.549132
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       47.000000
Name: count, dtype: float64

In [39]:
context_unique[context_unique['count']==1]

context  count
0    A few nights ago I talked to this girl I know ...      1
2    A friend of mine taking psychology advised I g...      1
3    A girl and I were madly in love. We dated for ...      1
8    About a month ago, I went through my boyfriend...      1
13   After a domestic abuse situation, I went to a ...      1
..                                                 ...    ...
830  When having sex I think of other men and want ...      1
837  Whenever I leave my girlfriend I get panic att...      1
839  Whenever I sing along with a song that I like ...      1
841                   Why am I attracted to older men?      1
844  i was addicted to porn since the second grade....      1

[452 rows x 2 columns]

In [40]:
count_distribution = context_unique['count'].value_counts()

count_distribution

count
1     452
2     184
3      74
4      43
5      25
6      18
9      12
7      11
8       6
11      5
23      2
21      2
15      2
10      2
38      1
26      1
41      1
12      1
47      1
22      1
13      1
Name: count, dtype: int64

In [41]:
CONTEXT_UNIQUE_FILENAME='dataset/context_unique.csv'

context_unique.to_csv(CONTEXT_UNIQUE_FILENAME)

In [42]:
conversation_dataset.describe()

context  \
count                                                2072   
unique                                                845   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2072  
unique                                               2072  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

### Data is cleaned and ready for simulation

So we have 845 unique requests, with a mean of 2.45 responses for each, and 452 having one response, and max of 47 responses, but only 10 requests have more then 10 responses.

### Conclusions

**Maybe for simulation we should use only requests with 2 or more responses, so we have some variety in approved responses to compare with. If we take request with 2 or more we will have 393 requests, if we take with 3 or more - 209 requests. 209 may be enough for our simulation. And having only one response will make it possible for incorrect scoring of ChatGPT output. Because mental issues is delicate matter, so several different responses may serve well for specific case.**


## ChatGPT

We follow the standard way of using OpenAI API for reaching ChatGPT
The below code uses ```.env``` file in this directory, which is not included to the repository.
```.env```  which must contain:
```
OPENAI_API_KEY=...
```
This API key to be used to interact with OpenAI API

If you want to use this code, you must [get your own API key from OpenAI](https://platform.openai.com/api-keys)

Moreover you will need to make a payment to your ChatGPT account, to be able to use it.

This code is based on the [course by DeepLearning.AI](https://learn.deeplearning.ai/courses/langchain/lesson/2/models%2C-prompts-and-parsers)

In [43]:
# !pip install python-dotenv
# !pip install openai

In [44]:
import os
import openai

from dotenv import load_dotenv, find_dotenv
_ = load_dotenv(find_dotenv()) # read local .env file

In [45]:
GPT_MODEL='gpt-3.5-turbo'   
#
# Pricing
# Model	Input	Output
# gpt-3.5-turbo-0125	$0.0005 / 1K tokens	$0.0015 / 1K tokens
#
# Updated GPT 3.5 Turbo
# The latest GPT-3.5 Turbo model with higher accuracy at responding in requested formats 
# and a fix for a bug which caused a text encoding issue for non-English language function calls. 
# Returns a maximum of 4,096 output tokens.
#

In [46]:
client = openai.OpenAI(

    api_key=os.environ.get("OPENAI_API_KEY"),
)

In [47]:
chat_completion = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": "Hello, this is a test",
        }
    ],
    model=GPT_MODEL,
)

print(chat_completion.choices[0].message.content)


Hello! How can I assist you today?


In [48]:
import random

TEMPERATURE = 0.6 # default temperature for ChatGPT. Web ChatGPT uses 0.6 all the time. varies from 0 (same answers, no randomness) to 1 (much variety, much random)

# Our randomness constants:

USE_STARTUP_DIALOG = 0.25

# if we use startup dialog, we may have up to 4 preceding phrases.

DIALOG_LENGTH_WEIGHTS=[0.35,0.3,0.2,0.15]  # So len 1 will occur more frequently, len 4 - much less
DIALOG_LENGTH_OPTIONS=[2,4,6,8]

starting_messages = [
[
    {'role': 'user', 'content': "Hi"},
    {'role': 'assistant', 'content': "Hello! How can I assist you today?"},
    {'role': 'user', 'content': "Yes! How are you?"},
    {'role': 'assistant', 'content': "As an AI, I don't have feelings or emotions, but I'm here and ready to assist you with any questions or tasks you have. How can I help you today?"},
    {'role': 'user', 'content': "What do you do"},
    {'role': 'assistant', 'content': "I'm here to help with a wide range of tasks! You can ask me questions, seek information on various topics, brainstorm ideas, get assistance with writing or editing, play text-based games, generate creative content, and much more. Just let me know what you need help with, and I'll do my best to assist you!"},
    {'role': 'user', 'content': 'Yes'},
    {'role': 'assistant', 'content': "Great! If you have any specific questions or need assistance with something, feel free to ask. I'm here to help!"},
],
[
    {'role': 'user', 'content': "Hello!"},
    {'role': 'assistant', 'content': "Hello! How can I assist you today?"},
    {'role': 'user', 'content': "I don't know"},
    {'role': 'assistant', 'content': "That's perfectly fine! If you have any questions, need advice, want to discuss a topic, or just want to chat, feel free to let me know. I'm here to help with anything you need."},
    {'role': 'user', 'content': "Ok. I will think about it"},
    {'role': 'assistant', 'content': "Sure thing! Take your time to think, and whenever you're ready, feel free to reach out. I'll be here to assist you whenever you need."},
    {'role': 'user', 'content': "Tell me something"},
    {'role': 'assistant', 'content': "Sure! Did you know that the world's largest desert isn't the Sahara, but actually Antarctica? Despite being covered in ice, Antarctica is considered a desert because it receives very little precipitation, less than 2 inches (50 millimeters) per year in some areas. Its size and harsh climate make it the largest desert on Earth, even larger than the Sahara Desert."},         
],
[
    {'role': 'user', 'content': "Nice to see you"},
    {'role': 'assistant', 'content': "Nice to see you too! How can I assist you today?"},
    {'role': 'user', 'content': "What can you say?"},
    {'role': 'assistant', 'content': "As an AI language model, I can engage in conversation, answer questions, provide information on a wide range of topics, generate text based on prompts, offer suggestions, and more. Just let me know what you need assistance with or what you'd like to discuss!"},
    {'role': 'user', 'content': "Just tell me something"},
    {'role': 'assistant', 'content': "Sure! Did you know that honey never spoils? Archaeologists have found pots of honey in ancient Egyptian tombs that are over 3,000 years old and still perfectly edible. This remarkable longevity is due to the unique chemical composition of honey, which includes low water content and natural acidity, making it inhospitable to bacteria and microorganisms that cause food to spoil."},
    {'role': 'user', 'content': "Interesting"},
    {'role': 'assistant', 'content': "I'm glad you found it interesting! If you have any other topics you'd like to learn about or discuss, feel free to let me know. I'm here to help!"},  
],
[
    {'role': 'user', 'content': "How are things"},
    {'role': 'assistant', 'content': "As an AI, I don't have personal experiences, but I'm here and ready to assist you with any questions or tasks you have! How can I help you today?"},  
    {'role': 'user', 'content': "I am unsure"},
    {'role': 'assistant', 'content': "It's okay to feel unsure sometimes. If there's anything specific on your mind or any decision you're facing, feel free to share, and I'll do my best to assist you. Whether it's about making a choice, seeking advice, or just discussing your thoughts, I'm here to help."},  
    {'role': 'user', 'content': "yes talk to me"},
    {'role': 'assistant', 'content': "Of course! Is there anything particular on your mind that you'd like to talk about? It could be anything from personal concerns, questions about a topic, or just casual conversation. Feel free to share, and I'm here to listen and offer assistance or insights as best as I can."},  
    {'role': 'user', 'content': "i don't know"},
    {'role': 'assistant', 'content': "That's completely fine. Sometimes it can be helpful to explore different topics or interests to see if anything sparks your curiosity or engages your thoughts. Is there a subject you've been curious about or something you've been meaning to learn more about? Alternatively, we can engage in a casual conversation about anything you'd like. Let me know if there's a specific direction you'd like to take our conversation or if there's anything specific you'd like to discuss."},  
],
]


### Helper function for request - response

In [49]:
def generate_messages(phrase_request:str)->list[dict]:
    """Using random this function will generate messages to pass to ChatGPT

    Args:
        phrase_request (str): Request phrase to put to the end of messages list

    Returns:
        list[dict]: messages list in a OpenAi format
    """
    # this is a standard option
    return_messages=[    {'role': 'system', 'content': "You are a helpful assistant."},
                     ]
    
    # use random to chose if we are adding anything
    if random.random()<USE_STARTUP_DIALOG:
        # choose number of adding, based on provided weights
        startup_len_to_add=random.choices(DIALOG_LENGTH_OPTIONS, weights=DIALOG_LENGTH_WEIGHTS)[0]
        message_chain=random.randint(0,len(starting_messages)-1)
        return_messages+=starting_messages[message_chain][:startup_len_to_add]
    
    # finally our phrase                                                                                
    return return_messages+[
        {'role': 'user', 'content': phrase_request},
    ]

In [50]:
generate_messages('hello')

[{'role': 'system', 'content': 'You are a helpful assistant.'},
 {'role': 'user', 'content': 'hello'}]

In [51]:
generate_messages('hello')

[{'role': 'system', 'content': 'You are a helpful assistant.'},
 {'role': 'user', 'content': 'Nice to see you'},
 {'role': 'assistant',
  'content': 'Nice to see you too! How can I assist you today?'},
 {'role': 'user', 'content': 'hello'}]

In [52]:
# now let's write a function that sends this messages to ChatGPT and get response

In [53]:
def request_gpt(messages:list[dict], temperature=TEMPERATURE, model=GPT_MODEL)->str:
    """Sends messages to ChatGPT and returns the response

    Args:
        messages (list[dict]): list of messages in OpenAI format

    Returns:
        str: response of ChatGpt
    """
    chat_completion = client.chat.completions.create(
        messages=messages,
        model=model,
        temperature=temperature
    )

    return chat_completion.choices[0].message.content

In [54]:
def speak_with_gpt(phrase_request:str)->str:
    """Use phrase_request to generate messages with generate_messages function, send's them to ChatGPT, and return response

    Args:
        phrase_request (str): Phrase to ask

    Returns:
        str: Response of ChatGPT
    """
    return request_gpt(generate_messages(phrase_request=phrase_request))

In [55]:
example_phrase=context_unique.loc[100]['context']

In [56]:
generate_messages(example_phrase)

[{'role': 'system', 'content': 'You are a helpful assistant.'},
 {'role': 'user',
  'content': 'I am always arguing with my father. He gets stressed over work and health and talks to me in a tone of voice that seems very demanding and seems more like yelling. I get upset often at this. Am I too sensitive? He always says I am overreacting but never seems to understand that he is hurting my feelings. No matter how often I try to tell him this, he never listens.'}]

In [57]:
# speak_with_gpt(example_phrase)

In [58]:
# lets try it again several times

In [59]:
# speak_with_gpt(example_phrase)

In [60]:
# speak_with_gpt(example_phrase)

In [61]:
# copies of output
chatgpt_response1="It's completely understandable that you feel upset when your father speaks to you in a demanding or yelling tone. It's important to remember that everyone has different communication styles and ways of expressing their stress. It's also okay to have your own feelings and reactions to the way he speaks to you.\n\nIt's not necessarily that you are too sensitive, but rather that you have valid feelings that deserve to be acknowledged and respected. It's important to communicate your feelings to your father in a calm and assertive manner. You can try to have a conversation with him during a calm moment, expressing how his tone of voice makes you feel and how you would like to be spoken to.\n\nIt's possible that your father may not be aware of how his tone affects you, and it may take some time for him to understand and adjust his communication style. It's important to be patient and persistent in expressing your feelings to him.\n\nIf despite your efforts, your father continues to dismiss your feelings and refuses to listen, it may be helpful to seek support from a trusted family member, friend, or counselor to help navigate the situation and find healthy ways to cope with the challenging interactions. Remember that it's okay to prioritize your own well-being and set boundaries in relationships that are emotionally challenging."
chatgpt_response2='It sounds like you\'re going through a challenging situation with your father. It\'s important to remember that everyone experiences and expresses stress differently. It\'s also common for misunderstandings to occur in communication, especially during times of tension.\n\nYour feelings are valid, and it\'s okay to feel upset when someone speaks to you in a way that hurts your feelings. It\'s also understandable that you want your father to understand how his tone affects you.\n\nHere are a few suggestions that may help improve communication with your father:\n\n1. Choose the right time: Try to have a calm and private conversation with your father when both of you are in a relaxed state and not already in the midst of an argument.\n\n2. Use "I" statements: Express your feelings using "I" statements to help your father understand how his tone of voice affects you. For example, "I feel hurt when you speak to me in a demanding tone."\n\n3. Listen actively: Try to listen to your father\'s perspective as well. Understanding each other\'s point of view can lead to better communication and empathy.\n\n4. Seek support: If you find it challenging to communicate with your father on your own, consider seeking the help of a family therapist or counselor who can facilitate productive conversations and provide guidance on improving your relationship.\n\nRemember, it\'s okay to set boundaries and communicate your feelings in a respectful manner. Your well-being is important, and it\'s essential to address any concerns that are affecting your mental health and relationships.'
chatgpt_response3="It's understandable that you feel upset in this situation. It's important to remember that your feelings are valid, and it's not necessarily about being too sensitive. It sounds like there may be a communication issue between you and your father.\n\nIt might be helpful to have a calm and honest conversation with your father when you both are in a good mood. Express how his tone of voice and demanding behavior make you feel, and try to explain that you are not trying to overreact but instead want to improve your relationship with him.\n\nIt's possible that your father may not realize the impact of his words and actions on you, so having an open and honest conversation could help bring this to his attention. It may take time for him to understand and change his behavior, so be patient and continue to communicate your feelings to him.\n\nIn the meantime, it could also be beneficial to find healthy ways to cope with the stress and tension in your relationship with your father, such as practicing self-care, talking to a trusted friend or counselor, or engaging in activities that bring you joy and relaxation. Remember to prioritize your well-being and set boundaries to protect your emotional health."


In [62]:
dataset_response=conversation_dataset[conversation_dataset['context']==example_phrase]['response'].iloc[0]
print(dataset_response)

Thanks for the question. Regardless of whether or not you are sensitive, it is okay to want better communication with those who are important to you. When we feel that we are being talked down to or being ridiculed, it is difficult to not take it personally. It sounds like you have taken the first and necessary step in resolving this by bringing up the issue to him. Is it possible that the delivery of your message is causing him to feel blamed or defensive? Try communicating with him during a time when you are not upset and when it is out of context. Begin your statements with "I" as opposed to "you." For example, you can tell him, "I feel angry when you raise your voice at me" instead of "You always yell at me." Also, avoid using black and white terms such as never and always. Additionally, offer sugggestions as to ways that he can better communicate with you. Maybe he just doesn't know how to. Lastly, when delivering a message, I like to use the sandwich method by starting off the di

### Helper function for comparing the results

In [63]:
COMPARING_TEMPERATURE=0   # lower is better in such cases, we will get more sustainable responses

#
# note: i have made several experiments with it. with high temp you sometimes have 20 f
#

COMPARING_PROMPT_TEMPLATE ="""
Please compare the meaning and summarization of PHRASE_1 and PHRASE_2. 

Rate similarity of meaning by scale from 0 to 100 (0 - opposite meaning, no similarity, 100 - total similarity).

Response with number only!

```PHRASE_1```
<<<PHRASE1###>>>
```END OF PHRASE_1```


```PHRASE_2```
<<<PHRASE2###>>>
```END OF PHRASE_2```

SCALE OF LIKENESS:
"""

In [64]:
def generate_comparing_prompt(response1:str,response2:str)->str:
    """Insert into COMPARING_PROMPT_TEMPLATE given request and responses

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        str: Prompt to send to ChatGRT for comparing
    """
    return COMPARING_PROMPT_TEMPLATE.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)

In [65]:
comparing_prompt=generate_comparing_prompt(chatgpt_response3,dataset_response)
print(comparing_prompt)


Please compare the meaning and summarization of PHRASE_1 and PHRASE_2. 

Rate similarity of meaning by scale from 0 to 100 (0 - opposite meaning, no similarity, 100 - total similarity).

Response with number only!

```PHRASE_1```
It's understandable that you feel upset in this situation. It's important to remember that your feelings are valid, and it's not necessarily about being too sensitive. It sounds like there may be a communication issue between you and your father.

It might be helpful to have a calm and honest conversation with your father when you both are in a good mood. Express how his tone of voice and demanding behavior make you feel, and try to explain that you are not trying to overreact but instead want to improve your relationship with him.

It's possible that your father may not realize the impact of his words and actions on you, so having an open and honest conversation could help bring this to his attention. It may take time for him to understand and change his beh

In [66]:
def compare_with_gpt(response1:str,response2:str, temperature=COMPARING_TEMPERATURE, model=GPT_MODEL)->int:
    """Makes a prompt to compare phrase1 and phrase2, send to ChatGPT and return and integer 

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        int: 0..100 - scale of likeness, -1 if something goes wrong
    """
    messages=[{'role': 'system', 'content': "You are a helpful assistant, proficient in psychology and language. "},
              {'role': 'user', 'content': generate_comparing_prompt(response1,response2)},  
    ]
    response=request_gpt(messages,temperature=temperature)
    
    try:
        scale_number=int(response.strip())
    except ValueError:
        scale_number=-1
        
    return scale_number
    

In [67]:
# print(compare_with_gpt(chatgpt_response1,dataset_response))

In [68]:
# print(compare_with_gpt(chatgpt_response2,dataset_response))

In [69]:
# print(compare_with_gpt("rtu7e56ujdfghj",dataset_response))

In [70]:
# print(compare_with_gpt(" YOU ARE BAD SON!!!",dataset_response))

In [80]:
example_phrase2=context_unique.iloc[341]['context']

In [72]:
example_phrase2

'I panicked over a minor parking lot mistake. It was totally my fault, but due to another insurance issue, I left a note with an illegible phone number. Guilty over this, I got an envelope containing several hundred dollars. Unfortunately, by the time I returned, the other driver had seen my ersatz note and drove away. Since I cannot make this up to that individual, how do I atone for my cowardly act? I have been trying to do random acts of kindness in hopes that karma will somehow even out.'

In [73]:
dataset_response=conversation_dataset[conversation_dataset['context']==example_phrase2]['response']
print(dataset_response)

407    From your writing, you have already atoned for...
408    Thanks for sharing as it shows your humanity a...
Name: response, dtype: object


In [74]:
dataset_response1=dataset_response.iloc[0]
dataset_response2=dataset_response.iloc[1]


In [75]:
print(dataset_response1,dataset_response2)

From your writing, you have already atoned for what you describe as "cowardly".Recognizing your action and trying to directly correct it, realizing your action was improper and demonstrating the willingness to take right action, are all examples of making things right.Since the actual driver is unavailable to consult as to their wishes on what they expect and feel would correct the matter, you did the next best thing and considered the situation from the driver point of view. Thanks for sharing as it shows your humanity and that you have a sensitive conscience. In your example here I'd encourage you to always think, "safety first"; and never offer a stranger at the accident scene "...several hundred dollars". Why? Because it could go badly against you (i.e. extortion, robbed, or would this be considered a bribe).   To the point, I'm reminded of Steps Eight and Nine in many 12-step programs regarding making amends. In your specific example since the injured unidentified party departed, 

In [76]:
chat_gpt_responses=[]
print("Example simulation.")
print("request:",example_phrase2)
print("database response1:", dataset_response1 )
print("database response2:", dataset_response2 )
print()

for i in range(4):
    print('iteration', i)
    single_response=speak_with_gpt(example_phrase2)
    print("GPT response 1:", single_response)
    comp_score1=compare_with_gpt(single_response,dataset_response1)
    comp_score2=compare_with_gpt(single_response,dataset_response2)
    print("Compared to dataset response 1", comp_score1)
    print("Compared to dataset response 2", comp_score2)   
    chat_gpt_responses.append([comp_score1,comp_score2,single_response])
    print()
    
print('final array of comparing:', chat_gpt_responses)

Example simulation.
request: I panicked over a minor parking lot mistake. It was totally my fault, but due to another insurance issue, I left a note with an illegible phone number. Guilty over this, I got an envelope containing several hundred dollars. Unfortunately, by the time I returned, the other driver had seen my ersatz note and drove away. Since I cannot make this up to that individual, how do I atone for my cowardly act? I have been trying to do random acts of kindness in hopes that karma will somehow even out.
database response1 From your writing, you have already atoned for what you describe as "cowardly".Recognizing your action and trying to directly correct it, realizing your action was improper and demonstrating the willingness to take right action, are all examples of making things right.Since the actual driver is unavailable to consult as to their wishes on what they expect and feel would correct the matter, you did the next best thing and considered the situation from t

In [77]:
chat_gpt_responses

[[85,
  85,
  "It sounds like you are feeling remorseful about the situation, and it's commendable that you are seeking ways to make amends. While you may not be able to directly rectify the specific incident with the other driver, your intention to balance things out through random acts of kindness is a positive step.\n\nHere are some suggestions on how you can continue to atone for the situation:\n\n1. **Continue with Random Acts of Kindness:** Keep spreading positivity and kindness in your daily interactions. Small gestures can make a significant impact on others' lives.\n\n2. **Reflect on the Experience:** Use this incident as a learning opportunity. Reflect on what happened, how it made you feel, and what you can do differently in the future to handle similar situations with more integrity and responsibility.\n\n3. **Practice Mindfulness and Self-Compassion:** Be kind to yourself as well. Acknowledge your mistake, learn from it, and forgive yourself. Practicing mindfulness can hel

In [88]:
example_phrase3=context_unique.iloc[1]['context']
dataset_response=conversation_dataset[conversation_dataset['context']==example_phrase3]['response']
print(example_phrase3)
print(dataset_response)

A few years ago I was making love to my wife when for no known reason I lost my erection,    Now I'm In my early 30s and my problem has become more and more frequent. This is causing major problems for my ego and it's diminishing my self esteem. This has resulted in ongoing depression and tearing apart my marriage.    I am devastated and cannot find a cause for these issues. I am very attracted to my wife and want to express it in the bedroom like I used to.    What could be causing this, and what can I do about it?
89    First step always is to do a medical rule out ...
90    When I'm working with men with this type of si...
91    For starters, know that this is a normal exper...
92    Erection maintenance is caused by connective t...
93    If you haven't already, please see a doctor.  ...
94    Assuming there are no medical explanations for...
Name: response, dtype: object


In [89]:
chat_gpt_responses2=[]
print("Example simulation 2.")
print("request:",example_phrase3)

print()

for i in range(5):
    print('iteration', i)
    single_response=speak_with_gpt(example_phrase3)
    print("Got GPT response", i)
    result_line=[]
    for j in range(6):
        result_line.append(compare_with_gpt(single_response,dataset_response.iloc[j])) 
    print('comparing scores:',result_line)
    result_line.append(single_response)
    chat_gpt_responses2.append(result_line)
    print()
    


Example simulation 2.
request: A few years ago I was making love to my wife when for no known reason I lost my erection,    Now I'm In my early 30s and my problem has become more and more frequent. This is causing major problems for my ego and it's diminishing my self esteem. This has resulted in ongoing depression and tearing apart my marriage.    I am devastated and cannot find a cause for these issues. I am very attracted to my wife and want to express it in the bedroom like I used to.    What could be causing this, and what can I do about it?

iteration 0
Got GPT response 0
comparing scores: [85, 85, 85, 20, 85, 85]

iteration 1
Got GPT response 1
comparing scores: [85, 85, 85, 72, 85, 85]

iteration 2
Got GPT response 2
comparing scores: [85, 85, 85, 72, 85, 85]

iteration 3
Got GPT response 3
comparing scores: [85, 85, 85, 72, 85, 85]

iteration 4
Got GPT response 4
comparing scores: [85, 85, 85, 25, 85, 85]



In [91]:
chat_gpt_responses2

[[85,
  85,
  85,
  20,
  85,
  85,
  "I'm sorry to hear that you're going through this challenging situation. Erectile dysfunction can be a complex issue with various potential causes, including physical, psychological, or a combination of both. It's essential to address this concern with the help of a healthcare professional, such as a doctor or a urologist, to determine the underlying cause and explore treatment options.\n\nSome common factors that can contribute to erectile dysfunction include stress, anxiety, relationship issues, underlying health conditions (such as heart disease, diabetes, or high blood pressure), hormonal imbalances, alcohol or substance abuse, medications, and lifestyle factors like smoking, lack of exercise, or poor diet.\n\nHere are some steps you can consider to address this issue:\n\n1. **Consult a Healthcare Professional**: Make an appointment with a doctor or urologist to discuss your concerns openly. They can help diagnose the underlying cause and recom

In [115]:
ii=38

example_phrase4=context_unique.iloc[ii]['context']
dataset_response_s4=conversation_dataset[conversation_dataset['context']==example_phrase4]['response']
print(example_phrase4)
print(dataset_response_s4)

Every time I send a message to someone or a group message on Instagram, iMessage, or snapchat people will read my messages but then they won't answer me.    Could it be that there is something they don't like about me? I don't understand why they won't answer my messages. How do I get people to respond to me?
1831    I understand the feeling of not being liked du...
1832    Sorry to hear your friends aren't responding t...
1833    I'm sorry that you are having a difficult time...
Name: response, dtype: object


In [117]:
chat_gpt_responses4=[]
print("Example simulation 3.")
print("request:",example_phrase4)

print()

for i in range(5):
    print('iteration', i)
    single_response=speak_with_gpt(example_phrase4)
    print("Got GPT response", i)
    result_line=[]
    for j in range(3):
        result_line.append(compare_with_gpt(single_response,dataset_response_s4.iloc[j])) 
    print('comparing scores:',result_line)
    result_line.append(single_response)
    chat_gpt_responses4.append(result_line)
    print()

Example simulation 3.
request: Every time I send a message to someone or a group message on Instagram, iMessage, or snapchat people will read my messages but then they won't answer me.    Could it be that there is something they don't like about me? I don't understand why they won't answer my messages. How do I get people to respond to me?

iteration 0
Got GPT response 0
comparing scores: [85, 72, 72]

iteration 1
Got GPT response 1
comparing scores: [85, 72, 85]

iteration 2
Got GPT response 2
comparing scores: [72, 85, 85]

iteration 3
Got GPT response 3
comparing scores: [85, 85, 85]

iteration 4
Got GPT response 4
comparing scores: [85, 85, 85]



In [118]:
chat_gpt_responses4

[[85,
  72,
  72,
  "I'm sorry to hear that you're feeling this way. It's important to remember that people's actions or lack of response to messages may not always be a reflection of who you are as a person. There could be various reasons why someone may not respond to messages promptly, such as being busy, forgetful, or simply not knowing how to respond at that moment.\n\nHere are a few tips that might help improve your chances of getting responses:\n\n1. **Be patient**: Give people some time to respond before assuming they are ignoring you. Remember, everyone has their own schedules and priorities.\n\n2. **Engage in meaningful conversations**: Try to initiate conversations that are interesting and engaging. Ask questions, share stories, and show genuine interest in the other person.\n\n3. **Avoid sending too many messages**: Sending too many messages in a short period of time may come across as overwhelming. Give people space to respond at their own pace.\n\n4. **Use different commu

In [119]:
for j in range(3):
    print(dataset_response_s4.iloc[j])

I understand the feeling of not being liked due to not having any responses on the sites you list.More than likely the non-response has less to do with liking you than w liking the way you write your messages.Have you tried asking a question in your posts?  This would be one way to invite people to answer your post.When you write,  imagine yourself as the reader of the post.If you think about the wording of a question which would motivate you to respond, then this formula will very likely be the same for many others who read your post.Good luck!
Sorry to hear your friends aren't responding to you. If these friends are in-person as well as online, perhaps going to them in person and asking talking to them about your concern. Perhaps it's the way you send messages or the way they are receiving them. if they say no, then a simple request to respond to your messages. Measure the result and notice if there are even small differences. Also understand, why this is so important to you. Ask you

## Help or advise needed

I am not sure in the correctness of my comparing. Results are very similar. On this point I will pause and ask for help, and maybe think myself. 

I am getting the 85 score most of the time

I think the comparing prompt may not be perfect. Need to think about it
